hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java

```diff
@@ -223,13 +233,56 @@ public Response put(
         }
       }
       throw ex;
-    } finally {
-      if (output != null) {
-        output.close();
-      }
     }
   }
 
+
+  private static void putKeyWithStream(OzoneBucket bucket,
+                                String keyPath,
+                                long length,
+                                int bufferSize,
+                                ReplicationConfig replicationConfig,
+                                Map<String, String> keyMetadata,
+                                InputStream body)
+      throws IOException {
+    try (OzoneDataStreamOutput streamOutput = bucket.createStreamKey(keyPath,
+            length, replicationConfig, keyMetadata)) {
+      writeToStreamOutput(streamOutput, body, bufferSize, length);
+    }
+  }
+
+  private static void writeToStreamOutput(OzoneDataStreamOutput streamOutput,
+                                          InputStream body,
+                                          int bufferSize)
+          throws IOException {
+    // (length == -1) Read up to EOF
+    writeToStreamOutput(streamOutput, body, bufferSize, -1);
+  }
+
+  private static void writeToStreamOutput(
+          OzoneDataStreamOutput streamOutput,
+          InputStream body,
+          int bufferSize,
+          long length) throws IOException {
+    byte[] buffer = new byte[bufferSize];
+    ByteBuffer writeByteBuffer;
+    long total = 0;
+    do {
+      int nn = body.read(buffer);
+      if (nn == -1) {
+        break;
+      } else if (nn != bufferSize) {
+        byte[] subBuffer = new byte[nn];
+        System.arraycopy(buffer, 0, subBuffer, 0, nn);
+        writeByteBuffer = ByteBuffer.wrap(subBuffer, 0, nn);
+      } else {
+        writeByteBuffer = ByteBuffer.wrap(buffer, 0, nn);
+      }
+      streamOutput.write(writeByteBuffer, 0, nn);
```

Comment:
We should avoid buffer copying.  I wonder if we should other body type (instead of InputStream) in order to support zero buffer copying; see https://docs.oracle.com/cd/E19798-01/821-1841/6nmq2cp2b/index.html#gkccg

Reply:
@szetszwo I looked at this document, Is it available `java.io.File All media types (*/*)`?

Reading files is similar to the implementation here: https://github.com/apache/ozone/blob/HDDS-4454/hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/keys/PutKeyHandler.java#L134

Reply:
@szetszwo I submitted the new code, is it ok to change it like this, please take a look.

Reply:
> ... Is it available java.io.File All media types (*/*)?

We should try it.

> Such a change is a little big, and the previous UT needs to be changed

I suggest we should change the case when uploadId == null (non-createMultipartKey) first.  Then, change the createMultipartKey case in a separate JIRA.

Reply:
> ...  all UT related to put cannot be used.

Could you give an example of why they cannot be reused?  We probably need to update the UTs.

Reply:
For example here

org.apache.hadoop.ozone.s3.endpoint.TestObjectPut

```
@Test
  public void testPutObject() throws IOException, OS3Exception {
    //GIVEN
    HttpHeaders headers = Mockito.mock(HttpHeaders.class);
    ByteArrayInputStream body =
        new ByteArrayInputStream(CONTENT.getBytes(UTF_8));
    objectEndpoint.setHeaders(headers);

    //WHEN
    Response response = objectEndpoint.put(bucketName, keyName, CONTENT
        .length(), 1, null, body);


    //THEN
    OzoneInputStream ozoneInputStream =
        clientStub.getObjectStore().getS3Bucket(bucketName)
            .readKey(keyName);
    String keyContent =
        IOUtils.toString(ozoneInputStream, UTF_8);

    Assert.assertEquals(200, response.getStatus());
    Assert.assertEquals(CONTENT, keyContent);
  }
```

Reply:
If streaming is enabled through configuration, do you need to retain the InputStream?

Reply:
We may add another put method as below:
```
+++ b/hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java
@@ -155,6 +153,24 @@ public Response put(
       @HeaderParam("Content-Length") long length,
       @QueryParam("partNumber")  int partNumber,
       @QueryParam("uploadId") @DefaultValue("") String uploadID,
+      @QueryParam("ozoneInternalParam_streaming") @DefaultValue("false")
+          boolean ozoneInternalParam_streaming,
+      File body) throws IOException, OS3Exception {
+    if (!ozoneInternalParam_streaming) {
+      final BufferedInputStream in = new BufferedInputStream(new FileInputStream(body));
+      return put(bucketName, keyPath, length, partNumber, uploadID, in);
+    }
+
+    // new code for Ratis streaming
+    ...
+  }
+
+  public Response put(
+      String bucketName,
+      String keyPath,
+      long length,
+      int partNumber,
+      String uploadID,
       InputStream body) throws IOException, OS3Exception {
 
     OzoneOutputStream output = null;
```

Reply:
Then, the existing tests will still work.  Of course, we need to create new tests for the new code.

Reply:
> 

Okay, I'm going to implement a version, and you can take a look